src/jtag/drivers/ftdi.c

```diff
@@ -711,6 +796,230 @@ static int ftdi_quit(void)
+	      }
+
+	    // drive another TCK without driving TMSC (TDO cycle)
+	    mpsse_clock_tms_cs(mpsse_ctx, &zero, 0, in, in_offset+i, 1, false, mode);
+	    
+  
+	  }
+}
+
+
+static void oscan1_mpsse_clock_tms_cs(struct mpsse_ctx *ctx, const uint8_t *out, unsigned out_offset, uint8_t *in,
+		       unsigned in_offset, unsigned length, bool tdi, uint8_t mode)
+{
+	uint8_t zero = 0;
+	uint8_t one = 1;
+
+	DEBUG_IO("oscan1_mpsse_clock_tms_cs: %sout %d bits, tdi=%d", in ? "in" : "", length, tdi);	
+  
+	for (unsigned i = 0; i < length; i++)
+	  {
+	    int bitnum;
+	    uint8_t tmsbit;
+	    uint8_t tdibit;
+
+	    // OSCAN1 uses 3 separate clocks
+
+	    // drive TMSC to the *negation* of the desired TDI value
+	    tdibit = tdi ? 0 : 1;
+
+	    // drive TMSC to desired TMS value
+	    bitnum = out_offset + i;
+	    tmsbit = ((out[bitnum/8] >> (bitnum%8)) & 0x1);
+
+	    if (tdibit == tmsbit)
+	      {
+		// Can squash into a single MPSSE command
+		uint8_t tmsbits = 0x3;
+		mpsse_clock_tms_cs_out(mpsse_ctx, &tmsbits, 0, 2, tdibit, mode);
+	      }
+	    else
+	      {
+		// Unoptimized case, can't formulate with a single command
+		mpsse_clock_tms_cs_out(mpsse_ctx, &one, 0, 1, tdibit, mode);		
+		mpsse_clock_tms_cs_out(mpsse_ctx, &one, 0, 1, (tmsbit != 0), mode);
+	      }
+
+	    // drive another TCK without driving TMSC (TDO cycle)
+	    mpsse_clock_tms_cs(mpsse_ctx, &zero, 0, in, in_offset+i, 1, false, mode);	    
+	  }
+}
+
+
+static void oscan1_mpsse_clock_tms_cs_out(struct mpsse_ctx *ctx, const uint8_t *out, unsigned out_offset,
+			   unsigned length, bool tdi, uint8_t mode)
+{
+	oscan1_mpsse_clock_tms_cs(ctx, out, out_offset, 0, 0, length, tdi, mode);
+}
+
+
+static void oscan1_set_tck_tms_tdi(struct signal *tck, char tckvalue, struct signal *tms, char tmsvalue, struct signal *tdi, char tdivalue)
+{
+	ftdi_set_signal(tms, tmsvalue);
+	ftdi_set_signal(tdi, tdivalue);  
+	ftdi_set_signal(tck, tckvalue);         
+}
+
+static void oscan1_reset_online_activate(void)
+{
+	struct signal *tck = find_signal_by_name("TCK");
+	struct signal *tdi = find_signal_by_name("TDI");
+	struct signal *tms = find_signal_by_name("TMS");
+	struct signal *tdo = find_signal_by_name("TDO");
+	uint16_t tdovalue;
+
+	if (!oscan1_mode)
+		return;
+	
+	/* After TAP reset, the OSCAN1-to-JTAG adapter is in offline and
+	non-activated state.  Escape sequences are needed to bring
+	the TAP online and activated into OSCAN1 mode. */
+
+	if (!tck)
+		LOG_ERROR("Can't run cJTAG online/activate escape sequences: TCK signal is not defined");
+
+	if (!tdi)
+		LOG_ERROR("Can't run cJTAG online/activate escape sequences: TDI signal is not defined");
+
+	if (!tms)
+		LOG_ERROR("Can't run cJTAG online/activate escape sequences: TMS signal is not defined");
+
+	if (!tdo)
+		LOG_ERROR("Can't run cJTAG online/activate escape sequences: TDO signal is not defined");
+	
+
+	/* TCK=0, TMS=1, TDI=0 (drive TMSC to 0 baseline) */
+	oscan1_set_tck_tms_tdi(tck,'0',tms,'1',tdi,'0');
```

Comment:
I love the comments here, but I think the compiled code would be much smaller if you had something like:
sequence[] = {{1,1,0},
{1,1,1},
...};

and then a loop to iterate over the sequence and call oscan1_set_tck_tms_tdi for each triple.

But I don't feel strongly, and I'm not convinced upstream would care too much either.